controllers/upgrade_controller.go

```diff
@@ -0,0 +1,270 @@
+	result, err := r.reconcilePhases()
+	if err != nil {
+		// TODO (lgangava): we are being very restrictive as of now, not upgrading if we hit any error
+		// differentiate between reconciler errors and not upgradeable errors
+		conditionStatus = metav1.ConditionFalse
+		conditionMessage = fmt.Sprintf("Operator could not be upgraded due to: %v", err)
+		r.log.Error(err, "an error occurred during reconcilePhases")
+	}
+	err = errors.Join(err, r.setOperatorCondition(conditionStatus, conditionMessage))
+
+	r.log.Info("reconciling completed")
+	return result, err
+}
+
+func (r *UpgradeReconciler) reconcilePhases() (ctrl.Result, error) {
+	if err := r.reconcileSubscriptionChannel(); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *UpgradeReconciler) reconcileSubscriptionChannel() error {
+	storageClientList := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClientList); err != nil {
+		return fmt.Errorf("failed to list storageclients %v", err)
+	}
+
+	if len(storageClientList.Items) == 0 {
+		r.log.Info("no storageclients exist")
+		return nil
+	}
+
+	desiredVersions := make([]string, 0, len(storageClientList.Items))
+	for idx := range storageClientList.Items {
+		desiredVersion := storageClientList.Items[idx].Status.Operator.DesiredVersion
+		// usually these versions will not have patch version, so we are adding an arbitrary num for semver comparison
+		desiredVersion += dummyPatchVersion
+		desiredVersions = append(desiredVersions, desiredVersion)
+	}
+
+	desiredVersion, err := getOldestVersion(desiredVersions)
+	if err != nil {
+		return fmt.Errorf("unable to find oldest desiredVersion that operator should be in: %v", err)
+	}
+
+	subscriptionList := &opv1a1.SubscriptionList{}
+	if err = r.list(subscriptionList, client.InNamespace(r.OperatorNamespace)); err != nil {
+		return fmt.Errorf("failed to list subscriptions in namespace %q: %v", r.OperatorNamespace, err)
+	}
+
+	ocsClientOperatorSubscription := utils.Find(subscriptionList.Items, func(sub *opv1a1.Subscription) bool {
+		return sub.Spec.Package == ocsClientOperatorSubscriptionPackageName
+	})
+
+	if ocsClientOperatorSubscription == nil {
+		return fmt.Errorf("unable to find subscription matching the package name %q", ocsClientOperatorSubscriptionPackageName)
+	}
+
+	channel := ocsClientOperatorSubscription.Spec.Channel
+	channelVersion, err := getVersionFromChannel(channel)
+	if err != nil {
+		return fmt.Errorf("failed to convert channel name %q to semver %v", channel, err)
+	}
+
+	// decide whether to patch the subscription or not
+	platformVersion, _ := semver.Make(r.PlatformVersion)
+	if desiredVersion.Major > platformVersion.Major || desiredVersion.Minor > platformVersion.Minor {
+		r.log.Info("backing off from updating subscription as operator version can become ahead of platform version")
+	} else if desiredVersion.Major == channelVersion.Major && desiredVersion.Minor == channelVersion.Minor+1 {
```

Comment:
Wondering is this the only possible scenario where the minor versions differ?

Reply:
No, this is the only scenario we are interested and wanted to take an action on. There's a corner case which I don't want to include which will complicate it.

4.20 -> released
4.21 -> released but whole minor version is faulty and redacted
4.22 -> released, now if desired is 22 but channel is at 20 this branch will not be taken

however, with an entire minor version being faulty we might have more serious issues that this.

do you mean any other scenarios?